### PR TITLE
Handle arrays in CFG generation

### PR DESCRIPTION
**Pull request by kamgal00**

### CODE REVIEW

File: app/src/main/kotlin/compiler/intermediate/ControlFlow.kt [hunk 1]
```diff
@@ -363,20 +405,94 @@ object ControlFlow {
...
+                is Expression.ArrayLength -> {
+                    val array = makeCFGForSubtree(astNode.expression)
+                    val resultReg = Register()
+                    cfgBuilder.addNextTree(
+                        IFTNode.RegisterWrite(
+                            resultReg,
+                            IFTNode.MemoryRead(IFTNode.Subtract(array, IFTNode.Const(memoryUnitSize.toLong())))
+                        )
+                    )
+                    cfgBuilder.addNextCFG(arrayMemoryManagement.genRefCountDecrement(array, expressionTypes[Ref(astNode.expression)]!!))
```
Comment (ignacy123) [inline app/src/main/kotlin/compiler/intermediate/ControlFlow.kt:417]: Where is the corresponding increment?

File: app/src/main/kotlin/compiler/intermediate/ControlFlow.kt [hunk 2]
```diff
@@ -363,20 +405,94 @@ object ControlFlow {
...
+                    IFTNode.RegisterRead(resultReg)
+                }
+                is Expression.ArrayElement -> {
+                    val array = makeCFGForSubtree(astNode.expression)
+                    val index = makeCFGForSubtree(astNode.index)
+                    val elementAddress =
+                        IFTNode.Add(array, IFTNode.Multiply(index, IFTNode.Const(memoryUnitSize.toLong())))
```
Comment (ignacy123) [inline app/src/main/kotlin/compiler/intermediate/ControlFlow.kt:424]: No need to split the line here, I think :thinking:

File: stdlib/stdlib.c [hunk 3]
```diff
@@ -27,6 +27,19 @@ void* _$checked_malloc(size_t size) {
     return address;
 }
 
+void _$populate_dynamic_array(uint64_t* address, uint64_t value, int64_t should_increment_elements) {
```
Comment (ignacy123) [inline stdlib/stdlib.c:30]:
```suggestion
void _$populate_dynamic_array(uint64_t* address, uint64_t value, int64_t should_increment_refcount) {
```

File: app/src/main/kotlin/compiler/intermediate/ControlFlow.kt [hunk 4]
```diff
@@ -363,20 +405,94 @@ object ControlFlow {
...
-            cfgBuilder.addNextTree(variableAccessGenerators[Ref(owner)]!!.genWrite(targetVariable, result, owner === currentFunction))
-        } else if (accessNodeConsumer == null) {
-            cfgBuilder.addNextTree(result)
+        if (target != null) {
+            when (target) {
+                is AssignmentTarget.VariableTarget -> {
+                    val owner = variableProperties[Ref(target.variable)]!!.owner
+                    cfgBuilder.addNextTree(variableAccessGenerators[Ref(owner)]!!.genWrite(target.variable, result, owner === currentFunction))
+                }
+                is AssignmentTarget.ArrayElementTarget -> {
```
Comment (ignacy123) [inline app/src/main/kotlin/compiler/intermediate/ControlFlow.kt:468]: Why aren't we increasing the refcount here? The assigned variable will be a new reference to the array, right?

File: app/src/test/kotlin/compiler/intermediate/ExpressionControlFlowTest.kt [hunk 5]
```diff
@@ -673,4 +690,83 @@ class ExpressionControlFlowTest {
                 )
             )
     }
+
+    private fun Int.toLiteral() = Expression.NumberLiteral(this.toLong())
+    private fun Int.toConst() = IFTNode.Const(this.toLong())
+
+    @Test
+    fun `simple array`() {
```
Comment (ignacy123) [inline app/src/test/kotlin/compiler/intermediate/ExpressionControlFlowTest.kt:698]: Some code snippets representing the tested operation would be nice (something similar to the tests above).

**Review by ignacy123 — CHANGES_REQUESTED** [review]

File: app/src/main/kotlin/compiler/intermediate/ControlFlow.kt (hunk 1 above)
Comment (kamgal00) [reply app/src/main/kotlin/compiler/intermediate/ControlFlow.kt:438]: It is done while evaluating `astNode.expression` to obtain `array` in the first line of the function.

File: app/src/main/kotlin/compiler/intermediate/ControlFlow.kt (hunk 4 above)
Comment (kamgal00) [reply app/src/main/kotlin/compiler/intermediate/ControlFlow.kt:502]: The same reason as above

**Pull request merged**